### PR TITLE
defaults 

### PR DESCRIPTION
**Pull request by phlptp**
Allow specification of default values for inputs in the json and toml configurations

### Summary

If merged this pull request will 
- add detection of "default" parameter
- update documentation regarding the "default" parameter

### CODE REVIEW

**Review by trevorhardy — CHANGES_REQUESTED** [review]
Let me do a minor edit on the Config. Options. Reference before you merge.

File: docs/references/configuration_options_reference.md [hunk 1]
```diff
@@ -761,6 +763,34 @@ This option allows the specific numerical definition of "change" when using the
...
 
+### `default`
+
+_API:_ `helicsInputSetDefaultX`
+[C++ input](https://docs.helics.org/en/latest/doxygen/classhelics_1_1Input.html#a55056ac9dd2895270f575827dd9951c7)
+| [C input](api-reference/C_API.md#input)
+| [Python input](https://python.helics.org/api/capi-py.html#helicsInputSetDefault)
+| [Julia input](https://julia.helics.org/latest/api/#HELICS.helicsInputSetDefulat-Tuple{HELICS.Input,XXX})
+
+This option allows the specific of the default value used when no publication has been received. Each datatype has its own API call such as:
```
Comment (nightlark) [inline docs/references/configuration_options_reference.md:774]:
```suggestion
This option allows specifying the default value used when no publication has been received. Each datatype has its own API call such as:
```

File: docs/references/configuration_options_reference.md [hunk 2]
```diff
@@ -761,6 +763,34 @@ This option allows the specific numerical definition of "change" when using the
...
+- `helicsInputSetDefaultChar()`
+- `helicsInputSetDefaultComplex()`
+- `helicsInputSetDefaultComplexVector()`
+- `helicsInputSetDefaulDouble()`
+- `helicsInputSetDefaultInteger()`
+- `helicsInputSetDefaultRaw()`
+- `helicsInputSetDefaultString()` (used for JSONs)
+- `helicsInputSetDefaultVector()`
+
+Note that the following, though they have syntaxical similarities to the above APIs, **do not** set default values on HELICS inputs when no publications have been received.
```
Comment (nightlark) [inline docs/references/configuration_options_reference.md:787]:
```suggestion
Note that the following, though they have syntactical similarities to the above APIs, **do not** set default values on HELICS inputs when no publications have been received.
```

File: docs/user-guide/fundamental_topics/interface_configuration.md [hunk 3]
```diff
@@ -93,15 +109,29 @@ Below is a sample JSON configuration file with some of the more common options.
...
 - **`subscriptions`** - These are lists of the values being sent to and from the given federate.
+
   - `key` - This string identifies the federation-unique value that this federate wishes to receive. If `global` has been set to `false` in the `publications` JSON configuration file, the name of the value is formatted as `<federate name>/<publication key>`. Both of these strings can be found in the publishing federate's JSON configuration file as the `name` and `key` strings, respectively. If `global` is `true` the string is the publishing federate's `key` value.
   - `required` - The message being subscribed to must be provided by some other publisher in the federation.
   - `type` - Data type, such as integer, double, complex.
   - `units` - Same as with `publications`.
   - `global` - Applies to the `key`, same as with `publications`.
+  - `default` - set the default value to return if not publications have been received
```
Comment (nightlark) [inline docs/user-guide/fundamental_topics/interface_configuration.md:121]:
```suggestion
  - `default` - set the default value to return if no publications have been received
```

File: docs/references/configuration_options_reference.md [hunk 4]
```diff
@@ -761,6 +763,34 @@ This option allows the specific numerical definition of "change" when using the
...
+- `helicsInputSetDefaulDouble()`
+- `helicsInputSetDefaultInteger()`
+- `helicsInputSetDefaultRaw()`
+- `helicsInputSetDefaultString()` (used for JSONs)
+- `helicsInputSetDefaultVector()`
+
+Note that the following, though they have syntactical similarities to the above APIs, **do not** set default values on HELICS inputs when no publications have been received.
+
+- `helicsInputSetDefaultTime()`
+- `helicsInputSetDefaultNamedPoint()`
```
Comment (phlptp) [inline docs/references/configuration_options_reference.md:790]: Are these not working,  this should do exactly the same thing as the others?  What makes you think they don't?

Comment (trevorhardy) [reply docs/references/configuration_options_reference.md:790]: As far as I know they are working fine they just aren't used to set the default value for an input that hasn't been published to.

**Comment by phlptp** [comment]
They are supposed to do the same thing as all the other functions in this list,.  If they are not working that is a bug and needs to be fixed.




---
Sent from Workspace ONE Boxer<https://whatisworkspaceone.com/boxer>

On October 17, 2022 at 1:24:56 PM PDT, Trevor Hardy ***@***.***> wrote:

@trevorhardy commented on this pull request.

________________________________

In docs/references/configuration_options_reference.md<https://urldefense.us/v3/__https://github.com/GMLC-TDC/HELICS/pull/2448*discussion_r997478166__;Iw!!G2kpM7uM-TzIFchu!m6NSzopeQRnSXNvHCMBgrd5HJij4x7JL3cRMIgwRnqxltZBxPuEw2qUGUd75z8c$>:

> +
+- `helicsInputSetDefaultBoolean()`
+- `helicsInputSetDefaultBytes()`
+- `helicsInputSetDefaultChar()`
+- `helicsInputSetDefaultComplex()`
+- `helicsInputSetDefaultComplexVector()`
+- `helicsInputSetDefaulDouble()`
+- `helicsInputSetDefaultInteger()`
+- `helicsInputSetDefaultRaw()`
+- `helicsInputSetDefaultString()` (used for JSONs)
+- `helicsInputSetDefaultVector()`
+
+Note that the following, though they have syntactical similarities to the above APIs, **do not** set default values on HELICS inputs when no publications have been received.
+
+- `helicsInputSetDefaultTime()`
+- `helicsInputSetDefaultNamedPoint()`


As far as I know they are working fine they just aren't used to set the default value for an input that hasn't been published to.

**Comment by phlptp** [comment]
Time translates to an int64 but takes a double

Named point is a string and double combined both are valid types



---
Sent from Workspace ONE Boxer<https://whatisworkspaceone.com/boxer>

On October 17, 2022 at 1:56:32 PM PDT, Trevor Hardy ***@***.***> wrote:

I think I'm confused about what those APIs are for. Are "time" and "named point" data types like "integer" and "vector"?

**Comment by trevorhardy** [comment]
OK. I'll modify that last paragraph to provide that clarification.

File: docs/references/configuration_options_reference.md [hunk 5]
```diff
@@ -761,6 +763,34 @@ This option allows the specific numerical definition of "change" when using the
...
+- `helicsInputSetDefaultComplexVector()`
+- `helicsInputSetDefaulDouble()`
+- `helicsInputSetDefaultInteger()`
+- `helicsInputSetDefaultRaw()`
+- `helicsInputSetDefaultString()` (used for JSONs)
+- `helicsInputSetDefaultVector()`
+
+Though they are not as obviously named, the following two APIs do provide a means of setting the default value as well:
+
+- `helicsInputSetDefaultTime()` - int64 but can also take a double
```
Comment (phlptp) [inline docs/references/configuration_options_reference.md:789]:
```suggestion
- `helicsInputSetDefaultTime()` - set the default using a HelicsTime value
```

File: docs/references/configuration_options_reference.md [hunk 6]
```diff
@@ -761,6 +763,34 @@ This option allows the specific numerical definition of "change" when using the
...
+- `helicsInputSetDefaultNamedPoint()` - string or a double
```
Comment (phlptp) [inline docs/references/configuration_options_reference.md:790]:
```suggestion
- `helicsInputSetDefaultNamedPoint()` -set the default NamedPoint which is a pair of a string and double used for tagged values or set points
```

**Comment by phlptp** [comment]
@nightlark  any other issues with this PR?

**Review by nightlark — APPROVED** [review]

**Comment by nightlark** [comment]
No, those are the only things I noticed.

**Pull request merged**